### DISCUSSION TITLE
LDAP related errors when building from source and having instantclient-sdk formula installed

### DISCUSSION THREAD

**Discussion by aleksandrs-ledovskis** [Homebrew/discussions#2487] [Everyday usage] [answered]
As all Homebrew packages for macOS 10.14 has now been pruned of pre-compiled bottles, any usage of `brew upgrade` incurs hefty recompilations.

What I have faced with couple of packages is a LDAP related error, for example (for `gnupg`)

```
ldap-misc.c:68:10: error: use of undeclared identifier 'LDAP_AUTH_METHOD_NOT_SUPPORTED'
    case LDAP_AUTH_METHOD_NOT_SUPPORTED: ec=GPG_ERR_LDAP_UNSUPPORTED_AUTH;break;
         ^
ldap-misc.c:122:10: error: use of undeclared identifier 'LDAP_AFFECTS_MULTIPLE_DSAS'
    case LDAP_AFFECTS_MULTIPLE_DSAS: ec = GPG_ERR_LDAP_AFFECTS_MULT_DSAS; break;
```

**Comment by aleksandrs-ledovskis** [comment, marked as answer]
The culprit is not related to Homebrew, but in fact is due to install of Oracle Instant Client, which squatted the `/usr/local/include/ldap.h` header. Upon removal of this file the Homebrew upgrade errors ceased and builds were successful.

Solution 1: `$ mv /usr/local/include/ldap.h /usr/local/include/ldap.h.bak`
Solution 2: `$ brew unlink instantclient-sdk`

Thanks to [pyldap/pyldap#61 (comment)](https://github.com/pyldap/pyldap/issues/61#issuecomment-377914001) for hint. Hopefully this knowledge helps someone in similar setup & situation.

**Reply by carlocab** [reply]
`brew doctor` should also warn you about the existence of an unbrewed `ldap.h` header.

**Reply by aleksandrs-ledovskis** [reply]
While `brew doctor` does show up header warnings, it seemingly doesn't catch the extraneous `ldap.h` header?

Here's the part of output that's relevant to headers when run on my machine:

```
Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/python3.5m/greenlet/greenlet.h
  /usr/local/include/python3.7m/greenlet/greenlet.h
```

Obviously, I've put back the Oracle Instant Client's header before running the command

```
$ file /usr/local/include/ldap.h
/usr/local/include/ldap.h: Algol 68 source text, ASCII text
```

**Reply by aleksandrs-ledovskis** [reply]
Ok, did check a bit more on why that `brew doctor` miss happens. Per `brew` [code](https://github.com/Homebrew/brew/blob/67cc3746f0fa8b27f9a48aa219e75bf2e4d4d9a8/Library/Homebrew/diagnostic.rb#L211), the only deterministic signal for stray files is "is file a symlink?", which mine `ldap.h` passes due to fact that it was populated by `homebrew-instantclient` tap's [`instantclient-sdk`](https://github.com/InstantClientTap/homebrew-instantclient/blob/eba8e478f942ca8edbcf44237d0a5570d3df3e0d/Formula/instantclient-sdk.rb) formula.

I don't blame Homebrew proper for this trouble, it really seems to boils down to how Oracle in its infinite knowledge chooses library header names. 🤦‍♂️ 

/cc @kaukas

<details>
<summary>Listing of such headers from "sdk/include/"</summary>

```
sdk/include/
├── ldap.h
├── nzerror.h
├── nzt.h
├── occi.h
├── occiAQ.h
├── occiCommon.h
├── occiControl.h
├── occiData.h
├── occiObjects.h
├── oci.h
├── oci1.h
├── oci8dp.h
├── ociap.h
├── ociapr.h
├── ocidef.h
├── ocidem.h
├── ocidfn.h
├── ociextp.h
├── ocikpr.h
├── ociver.h
├── ocixmldb.h
├── ocixstream.h
├── odci.h
├── orastruc.h
├── oratypes.h
├── oraxml.h
├── oraxmlcg.h
├── oraxsd.h
├── ori.h
├── orid.h
├── orl.h
├── oro.h
├── ort.h
├── xa.h
├── xml.h
├── xmldav.h
├── xmldf.h
├── xmlerr.h
├── xmlev.h
├── xmlotn.h
├── xmlproc.h
├── xmlsch.h
├── xmlsoap.h
├── xmlurl.h
├── xmlxptr.h
├── xmlxsl.h
└── xmlxvm.h
```
</details>

**Reply by carlocab** [reply]
Ah, so it's installed by a formula. Yea, that we can't help too much. Glad you sorted it out though. Next time `brew unlink instantclient-sdk` should probably work too.

**Reply by kaukas** [reply]
Is there a way to improve the formula to avoid messing up lib? `ldap.h`, `xml.h`… Oh dear…

**Reply by carlocab** [reply]
You can ask them to install things into a subdirectory of `include` instead. Maybe `include/"instantclient"`? Installing headers that shadow common ones into `include` is very uncool.

**Reply by kaukas** [reply]
Is there any extra work to be done to link user scripts against those headers afterwards? At the moment linking to headers in lib works without any extra env vars.

**Reply by carlocab** [reply]
They either need to change their source files from
```c
#include "ldap.h"
```
to
```c
#include "instantclient/ldap.h"
```
or add `"-I$(brew --prefix)/include/instantclient"` to their `C[XX,PP]FLAGS`.
